src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,65 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
                     lt::Union{Function, AbstractVector{<:Function}}=isless,
                     by::Union{Function, AbstractVector{<:Function}}=identity,
                     rev::Union{Bool, AbstractVector{Bool}}=false,
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by)
     return sort!(df, _alg, ord)
 end
 
-Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm, o::Base.Sort.Ordering) =
+function Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm,
+                    o::Base.Sort.Ordering, checkunique::Bool=false)
+    if checkunique
+        if !allunique(df)
+            throw(ArgumentError("Non-unique elements found. Multiple orders " *
+                                "are valid."))
+        end
+    end
     permute!(df, _sortperm(df, a, o))
+end
+
+# Internal function that aids in uniqueness checks
+# Converts column selectors to indices and checks necessary conditions for uniqueness
+function _perform_uniqueness_checks(df, cols, lt, by)
+    if !(lt === isless && by === identity)
+        throw(ArgumentError("Passing either lt or by along with checkunique=" *
+                            "true is not supported."))
+    end
+    # Easiest case, cols contains numeric indexes already
+    if cols isa AbstractVector{<:ColumnIndex}
+        by_or_lt_set = false
+        col_idxs = cols
+    # Second easiest, multicol index (no vector with orders clauses mixed in)
+    elseif cols isa MultiColumnIndex && !(cols isa AbstractVector) || cols isa ColumnIndex
+        by_or_lt_set = false
+        col_idxs = index(df)[cols]
+    elseif cols isa UserColOrdering
+        by_or_lt_set = haskey(cols.kwargs, :lt) || haskey(cols.kwargs, :by)
+        col_idxs = index(df)[_getcol(cols)]
+    # Mix of ColOrdering and other ColumnSelectors
+    elseif cols isa AbstractVector
+        newcols = Int[]
+        by_or_lt_set = false
+        for col in cols
+            if col isa UserColOrdering
+                by_or_lt_set = haskey(cols.kwargs, :lt) || haskey(cols.kwargs, :by)
+            end
+
+            push!(newcols, index(df)[_getcol(col)])
+        end
+        col_idxs = newcols
+    end
+    if by_or_lt_set
+        throw(ArgumentError("Order clauses with either by or lt set in combination " *
+                            "with checkunique=true are not supported"))
+    end
+    !allunique(df, col_idxs) &&
```

Comment:
```suggestion
    allunique(df, col_idxs) ||
```

Reply:
Sorry, I don't understand this suggestion. Why do we want to remove the negation and change the logic to or?

Reply:
because it is equivalent and simpler, see https://docs.julialang.org/en/v1/manual/control-flow/#Short-Circuit-Evaluation

Reply:
Thank you for the tip! 😄